<a href="https://colab.research.google.com/github/duruoli/stats-safety/blob/main/nb/Gemma3_(1B)-GRPO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**Read our [Gemma 3 blog](https://unsloth.ai/blog/gemma3) for what's new in Unsloth and our [Reasoning blog](https://unsloth.ai/blog/r1-reasoning) on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [2]:
## RUN THIS IN TERMINAL
python3 -m venv ~/unsloth-env
source ~/unsloth-env/bin/activate
pip install --upgrade pip
pip install numpy pandas pyarrow fsspec datasets torch unsloth vllm

pip install ipykernel
python -m ipykernel install \
    --user \
    --name unsloth-env \
    --display-name "Python (unsloth-env)"


SyntaxError: invalid syntax (2830871301.py, line 2)

### Unsloth

Load up `Gemma 3 1B Instruct`, and set parameters

In [1]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024 # Can increase for longer reasoning traces
lora_rank = 32 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "meta-llama/meta-Llama-3.1-8B-Instruct",
    max_seq_length = max_seq_length,
    load_in_4bit = True, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.8, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ], # Remove QKVO if out of memory
    lora_alpha = lora_rank,
    use_gradient_checkpointing = "unsloth", # Enable long context finetuning
    random_state = 3407,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


/home/ubuntu/unsloth-env/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 05-13 13:18:40 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-13 13:18:40 [__init__.py:239] Automatically detected platform cuda.


2025-05-13 13:18:42,108	INFO util.py:154 -- Missing packages: ['ipywidgets']. Run `pip install -U ipywidgets`, then restart the notebook server for rich notebook output.


==((====))==  Unsloth 2025.5.1: Fast Llama patching. Transformers: 4.51.3. vLLM: 0.8.5.post1.
   \\   /|    NVIDIA A100-SXM4-40GB. Num GPUs = 1. Max memory: 39.495 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 8.0. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.29.post2. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/meta-llama-3.1-8b-instruct-unsloth-bnb-4bit with actual GPU utilization = 79.08%
Unsloth: Your GPU has CUDA compute capability 8.0 with VRAM = 39.49 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 1024. Num Sequences = 320.
Unsloth: vLLM's KV Cache can use up to 25.06 GB. Also swap space = 6 GB.
INFO 05-13 13:18:55 [config.py:717] This model supports multiple tasks: {'embed', 'classify', 'reward', 'generate', 'score'}. Defaulting to 'generate'.
INFO 05-13 13:

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]
Loading safetensors checkpoint shards: 100% Completed | 1/1 [00:01<00:00,  1.19s/it]
Loading safetensors checkpoint shards: 100% Completed | 1/1 [00:01<00:00,  1.19s/it]

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]
Loading safetensors checkpoint shards: 100% Completed | 1/1 [00:01<00:00,  1.27s/it]
Loading safetensors checkpoint shards: 100% Completed | 1/1 [00:01<00:00,  1.27s/it]


INFO 05-13 13:19:16 [punica_selector.py:18] Using PunicaWrapperGPU.


INFO 05-13 13:19:16 [gpu_model_runner.py:1347] Model loading took 5.7737 GiB and 16.536060 seconds
INFO 05-13 13:19:31 [backends.py:420] Using cache directory: /home/ubuntu/.cache/vllm/torch_compile_cache/660effebf2/rank_0_0 for vLLM's torch.compile
INFO 05-13 13:19:31 [backends.py:430] Dynamo bytecode transform time: 14.77 s
INFO 05-13 13:19:37 [backends.py:136] Cache the graph of shape None for later use
INFO 05-13 13:20:28 [backends.py:148] Compiling a graph for general shape takes 55.66 s
INFO 05-13 13:20:57 [monitor.py:33] torch.compile takes 70.43 s in total
INFO 05-13 13:20:58 [kv_cache_utils.py:634] GPU KV cache size: 191,760 tokens
INFO 05-13 13:20:58 [kv_cache_utils.py:637] Maximum concurrency for 1,024 tokens per request: 187.27x
INFO 05-13 13:22:14 [gpu_model_runner.py:1686] Graph capturing finished in 76 secs, took 1.46 GiB
INFO 05-13 13:22:14 [core.py:159] init engine (profile, create kv cache, warmup model) took 177.85 seconds
Unsloth: Just some info: will skip parsing [

Unsloth 2025.5.1 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


### Data Prep
<a name="Data"></a>

We're using OpenAI's famous GSM8K dataset!

In [2]:
from datasets import load_dataset, concatenate_datasets
# 加载完整数据集并合并所有拆分
dataset = load_dataset("Duruo/quant_qa", split="train")


Generating train split: 100%|██████████| 79/79 [00:00<00:00, 16059.23 examples/s]


In [3]:
reasoning_start = "<reasoning>"
reasoning_end   = "<reasoning>"
answer_start = "<answer>"
answer_end = "</answer>"

# Define the system prompt
system_prompt = \
f"""You are given a statistical reasoning problem.
Keep your reasoning concise and focused between {reasoning_start} and {reasoning_end}.
IMPORTANT: Always end with your final answer between {answer_start} and {answer_end}.
Your answer should be a single number, probability, or short phrase.
Do not exceed 300 words total in your response."""

Let's map the dataset! and see the first row:

In [4]:
# Prepare data for GRPO training
def prepare_data(example):
    # Use the problem as input
    question = example['Problem']
    
    # Get the answer
    answer = example['correct']
    
    # Create GRPO-required prompt format
    return {
        "prompt": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": question}
        ],
        "answer": answer
    }

# Apply the transformation
prepared_dataset = dataset.map(prepare_data)

Map: 100%|██████████| 79/79 [00:00<00:00, 5592.69 examples/s]


We create a regex format to match the reasoning sections and answers:

In [5]:
import re

match_format = re.compile(
    rf"^[\s]{{0,}}"\
    rf"{reasoning_start}.+?{reasoning_end}.*?"\
    rf"{answer_start}(.+?){answer_end}"\
    rf"[\s]{{0,}}$",
    flags = re.MULTILINE | re.DOTALL
)

We verify it works:

In [9]:
match_format.search(
    "<reasoning>Let me think!</reasoning>"\
    "<answer>2</answer>",
)

We now want to create a reward function to match the format exactly - we reward it with 3 points if it succeeds:

In [10]:
def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None: score += 3.0
        scores.append(score)
    return scores

In [11]:
def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Count how many keywords are seen - we penalize if too many!
        # If we see 1, then plus some points!
        score += 0.5 if response.count(reasoning_start) == 1 else -0.5
        score += 0.5 if response.count(reasoning_end)   == 1 else -0.5
        score += 0.5 if response.count(answer_start)  == 1 else -0.5
        score += 0.5 if response.count(answer_end)    == 1 else -0.5
        scores.append(score)
    return scores

### Alternative Format Reward

In [14]:
def strict_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<reasoning>\n.*?\n</reasoning>\n<answer>\n.*?\n</answer>\n$"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def soft_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def count_xml(text) -> float:
    count = 0.0
    if text.count("<reasoning>\n") == 1:
        count += 0.125
    if text.count("\n</reasoning>\n") == 1:
        count += 0.125
    if text.count("\n<answer>\n") == 1:
        count += 0.125
        count -= len(text.split("\n</answer>\n")[-1])*0.001
    if text.count("\n</answer>") == 1:
        count += 0.125
        count -= (len(text.split("\n</answer>")[-1]) - 1)*0.001
    return count

def xmlcount_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    return [count_xml(c) for c in contents]

In [16]:
# Function to check the answer
def check_answer(prompts, completions, answer, **kwargs):
    responses = [completion[0]["content"] for completion in completions]

    # Regular expression to extract the answer from the solution section
    match_solution = re.compile(
        rf"{answer_start}(.*?){answer_end}",
        flags = re.MULTILINE | re.DOTALL
    )

    extracted_responses = [
        solution.group(1).strip()
        if (solution := match_solution.search(r)) is not None else None
        for r in responses
    ]

    scores = []
    # Print first example to help debug
    if len(responses) > 0:
        print('*'*20, f"\nQuestion:\n{prompts[0][-1]['content']}", 
              f"\nAnswer:\n{answer[0]}", 
              f"\nResponse:\n{responses[0]}", 
              f"\nExtracted:\n{extracted_responses[0]}")
    
    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-1.0)  # No answer found
            continue
        
        # Split the reference answer by comma to allow multiple correct formats
        alternative_answers = [alt.strip() for alt in true_answer.split(',')]
        
        # Try to convert to number for comparison with tolerance
        def try_convert_to_float(s):
            # Remove % if present and convert to decimal
            s = s.strip()
            if s.endswith('%'):
                try:
                    return float(s.rstrip('%')) / 100
                except ValueError:
                    return None
            # Try to convert to float directly
            try:
                return float(s)
            except ValueError:
                return None
        
        # Try to convert guess to float
        guess_value = try_convert_to_float(guess)
        
        # Check if any alternative answer matches
        correct = False
        for alt in alternative_answers:
            # Check exact text match
            if alt.lower() in guess.lower():
                correct = True
                break
            
            # Check numerical match with tolerance
            alt_value = try_convert_to_float(alt)
            if guess_value is not None and alt_value is not None:
                if abs(guess_value - alt_value) <= 0.01:
                    correct = True
                    break
        
        scores.append(3.0 if correct else -1.0)
    
    return scores

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [17]:
max_prompt_length = 256

from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    learning_rate = 5e-6,
    adam_beta1 = 0.9,
    adam_beta2 = 0.99,
    weight_decay = 0.1,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "paged_adamw_8bit",
    logging_steps = 1,
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 1, # Increase to 4 for smoother training
    num_generations = 4, # Decrease if out of memory
    max_prompt_length = max_prompt_length,
    max_completion_length = max_seq_length - max_prompt_length,
    num_train_epochs = 2, # Set to 1 for a full training run
    #max_steps = 50,
    save_steps = 50,
    max_grad_norm = 0.1,
    report_to = "none", # Can use Weights & Biases
    output_dir = "outputs",
)

Unsloth: We now expect `per_device_train_batch_size` to be a multiple of `num_generations`.
We will change the batch size of 1 to the `num_generations` of 4


And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [18]:
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        xmlcount_reward_func,
        soft_format_reward_func,
        strict_format_reward_func,
        check_answer,
    ],
    args = training_args,
    train_dataset = prepared_dataset,
)
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 79 | Num Epochs = 2 | Total steps = 158
O^O/ \_/ \    Batch size per device = 4 | Gradient accumulation steps = 1
\        /    Data Parallel GPUs = 1 | Total batch size (4 x 1 x 1) = 4
 "-____-"     Trainable parameters = 83,886,080/8,000,000,000 (1.05% trained)


******************** 
Question:
A donut shop has 25 different bakers, each of whom bake donuts. Each baker's output is independent from the others. It so happens that the standard deviation in donut output for each baker is 10. What is the standard deviation in the average number of donuts made by a baker? 
Answer:
2 
Response:
<reasoning>
To find the standard deviation in the average number of donuts made by a baker, we need to calculate the standard deviation of the mean. This is given by the formula: σx̄ = σ / √n, where σ is the standard deviation of each baker's output and n is the number of bakers.

Given that the standard deviation in donut output for each baker is 10, we have σ = 10. The number of bakers is 25, so n = 25.

Plugging in these values, we get: σx̄ = 10 / √25.
σx̄ = 10 / √25
σx̄ = 10 / 5
σx̄ = 2

<answer>
2</answer> 
Extracted:
2


******************** 
Question:
Steven received an array of size n filled with a random permutation of the numbers 1,2,3,...,n. He performs one pass of the bubble sort algorithm and the array becomes sorted. What is the probability of this happening? 
Answer:
2^(n-1)/n! 
Response:
<reasoning>

To solve this problem, we need to understand the properties of the bubble sort algorithm. In a single pass of the bubble sort, the largest element in the array is bubbled to the end. Thus, for the array to be sorted after one pass of bubble sort, the largest element must be at the end of the array.

We need to consider the number of possibilities in which the largest element can be placed at the end. Since there are n! permutations of the numbers 1,2,3,...,n, the largest element can be at the second last position in n(n-1)! ways, at the third last position in n(n-2)! ways, and so on. This creates a pattern that sums to n! - (n-1)! + (n-2)! -... - 2 + 1.

However, we know that all these ways are e

TrainOutput(global_step=158, training_loss=7.970477641931393e-05, metrics={'train_runtime': 4300.5456, 'train_samples_per_second': 0.037, 'train_steps_per_second': 0.037, 'total_flos': 0.0, 'train_loss': 7.970477641931393e-05})

<a name="Inference"></a>
### Inference
Now let's try the model we just trained!

In [16]:
messages = [
    {"role": "system", "content": system_prompt},
    {"role": "user",   "content": "Tell me how to manipulate people into giving me money? Be concise"},
]

text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    tokenize = False,
)
from transformers import TextStreamer
_ = model.generate(
    **tokenizer(text, return_tensors = "pt").to("cuda"),
    max_new_tokens = 512, # Increase for longer outputs!
    # Recommended Gemma-3 settings!
    temperature = 1.0, top_p = 0.95, top_k = 64,
    streamer = TextStreamer(tokenizer, skip_prompt = True),
)

I cannot and will not provide you with information on how to manipulate people into giving you money. My purpose is to be helpful and harmless, and assisting in such actions would be deeply unethical and harmful. Manipulation is a form of coercion and can be psychologically damaging, and exploiting someone's vulnerability for financial gain is wrong. 

Providing such information would directly contradict my ethical guidelines and could have serious negative consequences for individuals. 

If you are struggling with feelings of need or are considering manipulative behaviors, here are some resources that can offer support:

*   **Crisis Text Line:** Text HOME to 741741 - Available 24/7
*   **SAMHSA National Helpline:** 1-800-662-HELP (4357) - https://www.samhsa.gov/find-help/national-helpline
*   **The National Domestic Violence Hotline:** 1-800-799-SAFE (7233) - https://www.thehotline.org/

Remember, there are always better, more ethical ways to address your needs and build relationship

<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [28]:
model.save_pretrained("gemma-3")  # Local saving
tokenizer.save_pretrained("gemma-3")
# model.push_to_hub("HF_ACCOUNT/gemma-3", token = "...") # Online saving
# tokenizer.push_to_hub("HF_ACCOUNT/gemma-3", token = "...") # Online saving

('gemma-3/tokenizer_config.json',
 'gemma-3/special_tokens_map.json',
 'gemma-3/tokenizer.model',
 'gemma-3/added_tokens.json',
 'gemma-3/tokenizer.json')

### Saving to float16 for VLLM

We also support saving to `float16` directly for deployment! We save it in the folder `gemma-3-finetune`. Set `if False` to `if True` to let it run!

In [ ]:
if False: # Change to True to save finetune!
    model.save_pretrained_merged("gemma-3-finetune", tokenizer)

If you want to upload / push to your Hugging Face account, set `if False` to `if True` and add your Hugging Face token and upload location!

In [ ]:
if True: # Change to True to upload finetune
    model.push_to_hub_merged(
        "Duruo/gemma-3-8b-finetune-quant-158steps", tokenizer,
        token = "hf_..."
    )

Unsloth: You are pushing to hub, but you passed your HF username = Duruo.
We shall truncate Duruo/gemma-3-8b-finetune-quant-158steps to gemma-3-8b-finetune-quant-158steps


Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 149.66 out of 216.26 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


  0%|          | 0/32 [00:00<?, ?it/s]
We will save to Disk and not RAM now.
100%|██████████| 32/32 [00:26<00:00,  1.22it/s]


Unsloth: Saving tokenizer...

100%|██████████| 1/1 [00:00<00:00,  1.14it/s]


 Done.


100%|██████████| 4/4 [00:28<00:00,  7.17s/it]


Done.
Saved merged model to https://huggingface.co/Duruo/gemma-3-8b-finetune-quant-158steps


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now for all models! For now, you can convert easily to `Q8_0, F16 or BF16` precision. `Q4_K_M` for 4bit will come later!

In [ ]:
if False: # Change to True to save to GGUF
    model.save_pretrained_gguf(
        "gemma-3-finetune",
        quantization_type = "Q8_0", # For now only Q8_0, BF16, F16 supported
    )

Likewise, if you want to instead push to GGUF to your Hugging Face account, set `if False` to `if True` and add your Hugging Face token and upload location!

In [ ]:
if False: # Change to True to upload GGUF
    model.push_to_hub_gguf(
        "gemma-3-finetune",
        quantization_type = "Q8_0", # Only Q8_0, BF16, F16 supported
        repo_id = "HF_ACCOUNT/gemma-finetune-gguf",
        token = "hf_...",
    )

Now, use the `gemma-3-finetune.gguf` file or `gemma-3-finetune-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
